File: backend/src/routes/reports.js
import express from 'express';
import { body } from 'express-validator';
import authMiddleware from '../middlewares/auth.js';
import validateRequest from '../middlewares/validateRequest.js';
import { supabase } from '../services/supabaseClient.js';
const router = express.Router();
router.use(authMiddleware);
router.get('/', async (req, res, next) => {
    try {
        const userId = req.user.id;
        const { data, error } = await supabase.from('reports').select('*').eq('user_id', userId).order('generated_at', { ascending: false });
        if (error)
            return res.status(400).json({ error: error.message });
        res.json(data);
    }
    catch (err) {
        next(err);
    }
});
router.post('/export', [body('type').isIn(['monthly', 'yearly']).withMessage('Report type must be monthly or yearly.')], validateRequest, async (req, res, next) => {
    try {
        const userId = req.user.id;
        const { type } = req.body;
        const period = type === 'monthly' ? new Date().toLocaleString('default', { month: 'long', year: 'numeric' }) : new Date().getFullYear().toString();
        const { data, error } = await supabase.from('reports').insert([{ user_id: userId, report_type: `${type.charAt(0).toUpperCase() + type.slice(1)} Report`, period, metadata: { preparedAt: new Date().toISOString() }, download_url: null }]);
        if (error)
            return res.status(400).json({ error: error.message });
        res.status(201).json(data[0]);
    }
    catch (err) {
        next(err);
    }
});
export default router;

File: backend/routers/reports.py
from fastapi import APIRouter, Depends, HTTPException
import uuid
from datetime import datetime
from schemas import User, ReportExport
from dependencies import get_current_user
from mock_db import mock_storage

router = APIRouter()

@router.get("/")
def get_reports(user: User = Depends(get_current_user)):
    user_reports = [r for r in mock_storage["reports"] if r.get("user_id") == user.id]
    return sorted(user_reports, key=lambda x: x.get("generated_at", ""), reverse=True)

@router.post("/export", status_code=201)
def export_report(req: ReportExport, user: User = Depends(get_current_user)):
    if req.type not in ["monthly", "yearly"]:
        raise HTTPException(status_code=400, detail="Report type must be monthly or yearly.")
    
    now = datetime.utcnow()
    period = now.strftime("%B %Y") if req.type == "monthly" else str(now.year)
    
    report = {
        "id": str(uuid.uuid4()),
        "user_id": user.id,
        "report_type": f"{req.type.capitalize()} Report",
        "period": period,
        "metadata": {"preparedAt": now.isoformat()},
        "download_url": None,
        "generated_at": now.isoformat()
    }
    mock_storage["reports"].append(report)
    return report
